lib/internal/url.js

```diff
@@ -1453,18 +1455,29 @@ function pathToFileURL(filepath) {
     outURL.hostname = domainToASCII(hostname);
     outURL.pathname = encodePathChars(
       RegExpPrototypeSymbolReplace(backslashRegEx, StringPrototypeSlice(filepath, hostnameEndIndex), '/'));
-  } else {
-    let resolved = path.resolve(filepath);
-    // path.resolve strips trailing slashes so we must add them back
-    const filePathLast = StringPrototypeCharCodeAt(filepath,
-                                                   filepath.length - 1);
-    if ((filePathLast === CHAR_FORWARD_SLASH ||
-         (isWindows && filePathLast === CHAR_BACKWARD_SLASH)) &&
-        resolved[resolved.length - 1] !== path.sep)
-      resolved += '/';
-    outURL.pathname = encodePathChars(resolved);
-  }
-  return outURL;
+    return outURL;
+  }
+  let resolved = path.resolve(filepath);
+  // path.resolve strips trailing slashes so we must add them back
+  const filePathLast = StringPrototypeCharCodeAt(filepath,
+                                                 filepath.length - 1);
+  if ((filePathLast === CHAR_FORWARD_SLASH ||
+       (isWindows && filePathLast === CHAR_BACKWARD_SLASH)) &&
+      resolved[resolved.length - 1] !== path.sep)
+    resolved += '/';
+
+  // Call encodePathChars first to avoid encoding % again for ? and #.
+  resolved = encodePathChars(resolved);
+
+  // Question and hash character should be included in pathname.
+  // Therefore, encoding is required to eliminate parsing them in different states.
+  // This is done as an optimization to not creating a URL instance and
+  // later triggering pathname setter, which impacts performance
+  if (StringPrototypeIndexOf(resolved, '?') !== -1)
+    resolved = RegExpPrototypeSymbolReplace(questionRegex, resolved, '%3F');
+  if (StringPrototypeIndexOf(resolved, '#') !== -1)
+    resolved = RegExpPrototypeSymbolReplace(hashRegex, resolved, '%23');
+  return new URL(`file://${resolved}`);
```

Comment:
Did we measure if `new URL(resolved, 'file:')` was faster by any chance?

Reply:
I think it is not faster per se. The speedup comes from not setting `url.pathname`. Instead of `new URL()` + `url.pathname = pathname` there is only `new URL()` or at least this is my understanding.

Reply:
Having 2 parameters in URL constructor calls ada::parse method 2 times. Assuming, it takes around 150-250ns per URL, when having the ability to choose, it is always faster to use only 1 argument.

Reply:
Additionally URL constructor and pathname setter both crosses the C++ boundary. So reducing the usage of any of them, makes a lot of difference.